gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,450 @@
+			continue;
+
+		a = DatumGetArrayTypeP(str_datum);
+
+		if (is_db_tracked(a))
+		{
+			trackedDb = (tracked_db_t *) palloc0(sizeof(tracked_db_t));
+
+			trackedDb->dbid = DatumGetObjectId(oid_datum);
+			trackedDb->get_full_snapshot_on_recovery = full_snapshot_on_recovery(a);
+			dbids = lappend(dbids, trackedDb);
+		}
+	}
+
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	if (dbids != NIL)
+	{
+		ListCell   *cell;
+
+		if (dispatch_to_master)
+			dispatch_register_to_master(dbids);
+
+		dispatch_register_to_segments(dbids, uninitialized_segments);
+
+		foreach(cell, dbids)
+		{
+			pfree(lfirst(cell));
+		}
+
+		list_free(dbids);
+	}
+
+	LWLockAcquire(tf_shared_state->bloom_set.lock, LW_EXCLUSIVE);
+	tf_shared_state->bgworker_ready = true;
+	LWLockRelease(tf_shared_state->bloom_set.lock);
+}
+
+/* scan pg_db_role_setting, find all databases, bind blooms if necessary */
+static void
+arenadata_toolkit_worker(Datum main_arg)
+{
+	elog(LOG, "[arenadata toolkit] Starting background worker");
+
+	bool		master_initialized = false;
+
+	pqsignal(SIGHUP, tracking_sighup);
+	pqsignal(SIGTERM, tracking_sigterm);
+	pqsignal(SIGUSR1, tracking_sigusr1);
+
+	BackgroundWorkerUnblockSignals();
+
+	BackgroundWorkerInitializeConnection(DB_FOR_COMMON_ACCESS, NULL);
+
+	while (!got_sigterm)
+	{
+		int			rc;
+		List	   *uninitialized_segments = NIL;
+
+		CHECK_FOR_INTERRUPTS();
+
+		StartTransactionCommand();
+
+		if (extension_created())
+		{
+			elog(LOG, "[arenadata toolkit] Getting uninitialized segments");
+			uninitialized_segments = get_uninitialized_segments(uninitialized_segments);
+
+			if (!master_initialized || list_length(uninitialized_segments) > 0)
+			{
+				elog(LOG, "Dispatching register to segments");
+				dispatch_register(!master_initialized, uninitialized_segments);
+				list_free(uninitialized_segments);
+				uninitialized_segments = NIL;
+				master_initialized = true;
+			}
+		}
+		CommitTransactionCommand();
+
+		rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
+					   tracking_worker_naptime_sec * 1000);
+		ResetLatch(&MyProc->procLatch);
+
+		/* Emergency bailout if postmaster has died */
+		if (rc & WL_POSTMASTER_DEATH)
+		{
+			ereport(LOG, (errmsg("[arenadata toolkit] bgworker is being terminated by postmaster death.")));
+			proc_exit(1);
+		}
+
+		if (got_sighup)
+		{
+			got_sighup = false;
+		}
+	}
+
+	if (got_sigterm)
+		ereport(LOG, (errmsg("[arenadata toolkit] stop worker process")));
+
+	proc_exit(0);
+}
+
+void
+arenadata_toolkit_worker_register(void)
+{
+	worker.bgw_flags = BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+	worker.bgw_restart_time = BGW_DEFAULT_RESTART_INTERVAL;
+	worker.bgw_main = arenadata_toolkit_worker;
+	worker.bgw_notify_pid = 0;
+	worker.bgw_start_rule = NULL;
+	sprintf(worker.bgw_name, "arenadata_toolkit");
+
+	RegisterBackgroundWorker(&worker);
```

Comment:
To call the `RegisterBackgroundWorker` function, it is enough to use a LOCAL and NON-STATIC variable of the `BackgroundWorker` type.